### PR TITLE
Old S3 guide wrapper

### LINKED ISSUE #5311: `guide_gengrob()` was quietly removed from `ggplot2`'s namespace in version 3.4.2.9000
`guide_train()`, `guide_merge()`, `guide_geom()`, and `guide_gengrob()` were removed from the namespace of `ggplot2`'s development version I assume due to the guide system being ported to ggproto. These functions were added in version 3.0.0, and I would appreciate instead of quietly removing them in the new version, that these are kept in simply to throw an informative error.

for example instead of 

```r
guide_gengrob <- function(guide, theme) {
 UseMethod('guide_gengrob')
}
```

we have 

```r
guide_defunct <- function() {
  calling_fn <- parent.frame(2)
  calling_fn_env <- parent.env(calling_fn)

  rlang::abort(
    c(sprintf("`%s()` is defunct in favor of new ggproto `ggplot2::Guide` System",  deparse1(rlang::caller_call()[[1]])),
      'i' = sprintf('The package `%s` is using an old API which is not compatible with ggplot2 (v%s)',
                    calling_fn_env$.packageName,
                    packageVersion("ggplot2")),
      '*' = sprintf("Please contact the maintainer of `%s` to inform them of this change.",
                    calling_fn_env$.packageName)),
    call = calling_fn
  )
}

#' @export
guide_gengrob <- function(guide, theme) {
  guide_defunct()
}

#' @export
guide_train <- function(guide, scale, aesthetic = NULL) {
  guide_defunct()
}

#' @export
guide_merge <- function(guide, new_guide) {
  guide_defunct()
}

#' @export
guide_geom <- function(guide, layers, default_mapping) {
  guide_defunct()
}
```

It may be unlikely that the package that extends the old guide api will reach these functions first, but if they are called, then at least there is an informative message. 

Finally, in NEWS.md, the changes to these functions should be documented, even if these changes aren't implemented, we should at least state that they were removed.

This issue was brought to my attention by a user of my package jtlandis/ggside#46

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #5311.

Briefly, it introduces the `GuideOld` ggproto class that mostly calls the old S3 guide generics at the appropriate places, and throws warnings and errors in appropriate places.

Less briefly:
I've been giving some thought on how to minimise the fallout from rewriting the guide system from S3 to ggproto. In #5311, it was suggested to bring back the old S3 methods, which is the major reason why the development version of ggplot2 is currently incompatible with packages that implement guide extensions (or depend on such packages).  So this PR adds these back.

In addition, this PR also provides a fallback mechanism for guides that have reimplemented all the S3 methods in the shape of a `GuideOld` ggproto class. When ggplot2 encounters an old S3 guide, it tries to wrap this guide in a `GuideOld` class that calls the S3 generics on the list it receives. An example of a guide that implements all relevant methods is Thomas' `ggraph::guide_edge_direction()` or Claus' `multiscales::guide_colourfan()`. If such an old guide is provided, a deprecation warning is shown (note 3.5.0. is putative version number):

``` r
library(ggplot2)
packageVersion("ggplot2") # dev version with this PR
#> [1] '3.4.2.9000'
library(multiscales)

colors <- scales::colour_ramp(
  colors = c(red = "#AC202F", purple = "#740280", blue = "#2265A3")
)((0:7)/7)

ggplot(US_polling) + 
  geom_sf(aes(fill = zip(Clinton_lead, moe_normalized)), color = "gray30", size = 0.2) + 
  coord_sf(datum = NA) +
  bivariate_scale(
    "fill",
    pal_vsup(values = colors, max_desat = 0.8, pow_desat = 0.2, max_light = 0.7, pow_light = 1),
    name = c("Clinton lead", "uncertainty"),
    limits = list(c(-40, 40), c(0, 1)),
    breaks = list(c(-40, -20, 0, 20, 40), c(0, 0.25, 0.50, 0.75, 1.)),
    labels = list(waiver(), scales::percent),
    guide = "colourfan"
  )
#> Warning: The S3 guide system was deprecated in ggplot2 3.5.0.
#> ℹ It has been replaced by a ggproto system that can be extended.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
```

However, if a guide only partially implements the S3 methods, either because the method calls `NextMethod()` or attempts to sub-class an old guide, an error is thrown that these methods have been superseded. For example, my own `ggh4x::guide_stringlegend()` only has a `guide_gengrob()` method, and none of the other methods.

``` r
library(ggplot2)
packageVersion("ggplot2") # dev version
#> [1] '3.4.2.9000'
library(ggh4x)

ggplot(mpg, aes(displ, hwy, colour = class)) +
  geom_point() +
  guides(colour = guide_stringlegend())
#> Error in `guide_train()` at ggplot2/R/guide-old.R:92:4:
#> ! <Guide> classes have been rewritten as <ggproto> classes.
#> The old S3 guide methods have been superseded.
```

<sup>Created on 2023-06-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I think this strikes a decent balance between the old and the new system. It doesn't burden the ggplot2 developers with maintaining all the old S3 methods, and guide extension maintainers have options. They can choose to update to the new ggproto system, or reimplement the old S3 methods in their own package if they prefer to maintain backwards compatibility.

### CODE REVIEW

**Review by thomasp85 — APPROVED** [review]
LGTM - once merged and we have figured out which packages still breaks (due to subclassing old guides) I think you should create PRs for these packages to show how to fix it

**Comment by teunbrand** [comment]
Yeah that should be managable

**Pull request merged**
